File: tests/python/test_client.py
# SPDX-License-Identifier: PMPL-1.0-or-later
# SPDX-FileCopyrightText: 2024-2026 Hyperpolymath Contributors

"""Tests for Economic Toolkit Python client."""

import pytest
from datetime import date
from economic_toolkit import EconomicToolkit, FRED, WorldBank
from economic_toolkit.formulas import elasticity, gdp_growth, gini_coefficient


class TestEconomicToolkit:
    """Test main client functionality."""

    def test_init_rest_mode(self):
        """Test initialization in REST mode."""
        client = EconomicToolkit(mode="rest")
        assert client.mode == "rest"
        assert client.api_url == "http://localhost:8080"

    def test_init_with_api_key(self):
        """Test initialization with API key."""
        client = EconomicToolkit(api_key="test-key-123")
        assert "Authorization" in client.session.headers
        assert client.session.headers["Authorization"] == "Bearer test-key-123"

    def test_custom_api_url(self):
        """Test custom API URL."""
        client = EconomicToolkit(api_url="https://api.example.com/")
        assert client.api_url == "https://api.example.com"


class TestDataSources:
    """Test data source clients."""

    def test_fred_client(self):
        """Test FRED client initialization."""
        fred = FRED()
        assert fred.source_id == "fred"
        assert fred.source_name == "Federal Reserve Economic Data"

    def test_worldbank_client(self):
        """Test World Bank client initialization."""
        wb = WorldBank()
        assert wb.source_id == "worldbank"
        assert wb.source_name == "World Bank"


class TestFormulas:
    """Test economic formula functions."""

    def test_elasticity_point(self):
        """Test point elasticity calculation."""
        quantity = [100, 90, 80, 70]
        price = [10, 11, 12, 13]
        result = elasticity(quantity, price, method="point")
        assert isinstance(result, float)
        assert result < 0  # Demand curve slopes downward

    def test_elasticity_arc(self):
        """Test arc elasticity calculation."""
        quantity = [100, 80]
        price = [10, 12]
        result = elasticity(quantity, price, method="arc")
        assert isinstance(result, float)

    def test_gdp_growth(self):
        """Test GDP growth rate calculation."""
        gdp = [100, 102, 105, 108]
        growth = gdp_growth(gdp)
        assert len(growth) == 3
        assert all(g > 0 for g in growth)

    def test_gdp_growth_annualized(self):
        """Test annualized GDP growth rate."""
        gdp = [100, 110]
        growth = gdp_growth(gdp, periods=1)
        assert isinstance(growth, float)
        assert growth > 0

    def test_gini_coefficient_equal(self):
        """Test Gini coefficient with equal distribution."""
        incomes = [100, 100, 100, 100]
        gini = gini_coefficient(incomes)
        assert gini < 0.01  # Near perfect equality

    def test_gini_coefficient_unequal(self):
        """Test Gini coefficient with unequal distribution."""
        incomes = [10, 20, 50, 100, 500]
        gini = gini_coefficient(incomes)
        assert 0 < gini < 1


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: tests/julia/test_client.jl
# SPDX-License-Identifier: MPL-2.0
# SPDX-FileCopyrightText: 2024-2026 Hyperpolymath Contributors

"""
Tests for Economic Toolkit Julia client.

Run with: julia --project=. tests/julia/test_client.jl
"""

using Test

# Add src to load path so EconomicToolkit module is found
push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "src", "julia"))
using EconomicToolkit

@testset "EconomicToolkit Tests" begin

    @testset "EconomicClient" begin
        @testset "default initialization" begin
            client = EconomicClient()
            @test client.api_url == "http://localhost:8080"
            @test client.api_key === nothing
        end

        @testset "initialization with API key" begin
            client = EconomicClient(api_key="test-key-123")
            @test haskey(client.headers, "Authorization")
            @test client.headers["Authorization"] == "Bearer test-key-123"
        end

        @testset "custom API URL strips trailing slash" begin
            client = EconomicClient(api_url="https://api.example.com/")
            @test client.api_url == "https://api.example.com"
        end
    end

    @testset "DataSources" begin
        @testset "FRED client" begin
            fred = FRED()
            @test source_id(fred) == "fred"
            @test source_name(fred) == "Federal Reserve Economic Data"
        end

        @testset "WorldBank client" begin
            wb = WorldBank()
            @test source_id(wb) == "worldbank"
            @test source_name(wb) == "World Bank"
        end
    end

    @testset "Formulas" begin
        @testset "point elasticity" begin
            quantity = [100.0, 90.0, 80.0, 70.0]
            price = [10.0, 11.0, 12.0, 13.0]
            result = elasticity(quantity, price; method="point")
            @test result isa Float64
            @test result < 0  # Demand curve slopes downward
        end

        @testset "arc elasticity" begin
            quantity = [100.0, 80.0]
            price = [10.0, 12.0]
            result = elasticity(quantity, price; method="arc")
            @test result isa Float64
        end

        @testset "GDP growth period-over-period" begin
            gdp = [100.0, 102.0, 105.0, 108.0]
            growth = gdp_growth(gdp)
            @test length(growth) == 3
            @test all(g -> g > 0, growth)
        end

        @testset "GDP growth annualized" begin
            gdp = [100.0, 110.0]
            growth = gdp_growth(gdp; periods=1)
            @test growth isa Float64
            @test growth > 0
        end

        @testset "Gini coefficient equal distribution" begin
            incomes = [100.0, 100.0, 100.0, 100.0]
            gini = gini_coefficient(incomes)
            @test gini < 0.01  # Near perfect equality
        end

        @testset "Gini coefficient unequal distribution" begin
            incomes = [10.0, 20.0, 50.0, 100.0, 500.0]
            gini = gini_coefficient(incomes)
            @test 0 < gini < 1
        end

        @testset "Lorenz curve" begin
            incomes = [20000.0, 30000.0, 40000.0, 50000.0, 100000.0]
            pop, inc = lorenz_curve(incomes)
            @test pop[1] == 0.0
            @test inc[1] == 0.0
            @test pop[end] == 1.0
            @test inc[end] ≈ 1.0
            @test length(pop) == length(incomes) + 1
        end

        @testset "CAGR" begin
            result = cagr(100.0, 150.0, 5)
            @test result isa Float64
            @test result > 0
        end

        @testset "growth_rate" begin
            result = growth_rate([100.0, 105.0, 110.0])
            @test length(result) == 2
            @test all(r -> r > 0, result)
        end
    end
end
